### PR TITLE
Convert mod_articles_categories to new structure

### PR DESCRIPTION
**Pull request by carlitorweb**
### Summary of Changes
This PR converts mod_articles_categories to the new structure

### Testing Instructions
- Use Joomla 4.3
- Create an instance of the module Articles - Categories. Note how it works.
- Apply patch
- Delete administrator/cache/autoload_psr4.php
- Refresh the page, make sure the module still displays the same as before.

### Additional comments
I not sure of the best approach to use cache here, _if is needed_.

### CODE REVIEW

**Review by chmst — APPROVED** [review]

File: modules/mod_articles_categories/tmpl/default.php [hunk 1]
```diff
@@ -12,10 +12,16 @@
 
 use Joomla\CMS\Helper\ModuleHelper;
 
-if (!$list) {
+if (!$list || empty($list)) {
```
Comment (ReLater) [inline modules/mod_articles_categories/tmpl/default.php]:
```suggestion
if (empty($list)) {
```

Comment (carlitorweb) [reply modules/mod_articles_categories/tmpl/default.php]: 😞

**Comment by carlitorweb** [comment]
@Quy any hint why is failing here the drone?

File: modules/mod_articles_categories/src/Helper/ArticlesCategoriesHelper.php [hunk 2]
```diff
@@ -19,37 +26,73 @@
 /**
  * Helper for mod_articles_categories
  *
- * @since  1.5
+ * @since  __DEPLOY_VERSION__
```
Comment (laoneo) [inline modules/mod_articles_categories/src/Helper/ArticlesCategoriesHelper.php]: Revert

File: modules/mod_articles_categories/src/Helper/ArticlesCategoriesHelper.php [hunk 3]
```diff
@@ -19,37 +26,73 @@
...
-        $categories = Categories::getInstance('Content', $options);
-        $category   = $categories->get($params->get('parent', 'root'));
+        // Get the number of items in this category or
+        // descendants of this category at the expense of performance.
+        $options['countItems'] = $moduleParams->get('numitems', 0);
 
-        if ($category !== null) {
-            $items = $category->getChildren();
+        /** @var CategoryInterface $contentCategoryService */
+        $contentCategoryService = new Category($options);
```
Comment (laoneo) [inline modules/mod_articles_categories/src/Helper/ArticlesCategoriesHelper.php]: Boot here the content component as it implements https://github.com/joomla/joomla-cms/blob/4.4-dev/libraries/src/Categories/CategoryServiceInterface.php. Then get the category from there.

File: modules/mod_articles_categories/src/Helper/ArticlesCategoriesHelper.php [hunk 4]
```diff
@@ -19,37 +26,73 @@
...
-            $count = $params->get('count', 0);
+        /** @var CategoryNode $parentCategory */
+        $parentCategory = $contentCategoryService->get($moduleParams->get('parent', 'root'));
 
-            if ($count > 0 && \count($items) > $count) {
-                $items = \array_slice($items, 0, $count);
-            }
+        $childrenCategories = [];
+
+        if ($parentCategory !== null) {
```
Comment (laoneo) [inline modules/mod_articles_categories/src/Helper/ArticlesCategoriesHelper.php]: To make the code a bit easier to read and prevent nested if's. Invert here the condition and a return $childrenCategories inside the if block.

File: modules/mod_articles_categories/tmpl/default_items.php [hunk 5]
```diff
@@ -15,36 +15,42 @@
 use Joomla\CMS\Router\Route;
 use Joomla\Component\Content\Site\Helper\RouteHelper;
 
-$input  = $app->getInput();
-$option = $input->getCmd('option');
-$view   = $input->getCmd('view');
-$id     = $input->getInt('id');
+?>
```
Comment (laoneo) [inline modules/mod_articles_categories/tmpl/default_items.php]: Here the same, revert all the output changes and make them against 5.0.

File: modules/mod_articles_categories/src/Helper/ArticlesCategoriesHelper.php (hunk 4 above)
Comment (carlitorweb) [reply modules/mod_articles_categories/src/Helper/ArticlesCategoriesHelper.php]: Invert how? You need kown first if the parent category is not null for try then get the children categories. If not children, you need return a empty array.

Comment (laoneo) [reply modules/mod_articles_categories/src/Helper/ArticlesCategoriesHelper.php]: Something like
```
       if ($parentCategory !== null) {
           retrun [];
      }
        $childrenCategories = [];

        
            // Get all the children categories of this node
            $childrenCategories = $parentCategory->getChildren(true);

            $count = $moduleParams->get('count', 0);

            if ($count > 0 && \count($childrenCategories) > $count) {
                $childrenCategories = \array_slice($childrenCategories, 0, $count);
            }

        return $childrenCategories;
```

**Comment by carlitorweb** [comment]
@laoneo cy test added

**Comment by carlitorweb** [comment]
@laoneo I think in this module is not needed revert the Helper class, mainly here what is done is removing the `Categories::getInstance` deprecated function, and changed some variables name. 

But please, let me kown if you need I rever the whole Helpr class.

**Comment by laoneo** [comment]
Can you include the cache the same way as in #40316?

**Comment by drmenzelit** [comment]
I have tested this item :white_check_mark: successfully on 964bb0b65cecbb159d518e04d616e0403a9a5e1c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39552">issues.joomla.org/tracker/joomla-cms/39552</a>.</sub>

**Pull request merged**